File: leetcode/medium/139-wordBreak.java
// first pass solution
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> seen = new HashSet<>();
        LinkedList<String> queue = new LinkedList<>();
        
        queue.addFirst(s);
        
        while (!queue.isEmpty()) {
            String currPath = queue.removeFirst();
            
            if (!seen.contains(currPath)) {
                if (currPath.length() == 0) return true;
                
                seen.add(currPath);

                for (String word : wordDict) {
                    if (currPath.startsWith(word)) {
                        queue.addLast(currPath.substring(word.length()));
                    }
                }
            }
        }
        return false;
    }
}

// second pass solution
class Solution {
    Set<String> visited = new HashSet<>();
    boolean flag = false;
    
    public boolean wordBreak(String s, List<String> wordDict) {
        recurse(wordDict, s); 
        
        return flag;
    }
    
    public void recurse(List<String> wordDict, String currPath) {
        if (currPath.length() == 0) {
            flag = true;
            return;
        }
        
        if (visited.contains(currPath) || flag) return;
        
        visited.add(currPath);
        
        for (String word : wordDict) {
            if (currPath.startsWith(word)) {
                recurse(wordDict, currPath.substring(word.length()));
            }
        }
    }
}

File: leetcode/medium/139-wordBreak.js
// Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, determine if s can be segmented into a 
// space-separated sequence of one or more dictionary words.

// Note:

// The same word in the dictionary may be reused multiple times in the segmentation.
// You may assume the dictionary does not contain duplicate words.
// Example 1:

// Input: s = "leetcode", wordDict = ["leet", "code"]
// Output: true
// Explanation: Return true because "leetcode" can be segmented as "leet code".
// Example 2:

// Input: s = "applepenapple", wordDict = ["apple", "pen"]
// Output: true
// Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
//              Note that you are allowed to reuse a dictionary word.

// first pass iterative approach
var wordBreak = function(s, wordDict) {
    let seen = new Set();
    let queue = [s];
    
    while (queue.length > 0) {
        let curr = queue.shift();
        
        if (!seen.has(curr)) {
            if (curr.length === 0) return true;
        
            seen.add(curr);
            
            for (let word of wordDict) {
                if (curr.startsWith(word)) {
                    queue.push(curr.slice(word.length))
                }
            }
        }
    }
    return false;
};

// second pass recursive approach
var wordBreak = function(s, wordDict) {
    let seen = new Set();
    
    function helper(s) {
        if (s.length === 0) return true;
        
        if (!seen.has(s)) {
            seen.add(s);          
            for (let word of wordDict) {
                if (s.startsWith(word)) {
                    if (helper(s.slice(word.length))) {
                        return true;
                    }
                }
            }
        }
        return false
    }
    return helper(s);
};
